src/renderer/components/app.tsx

```diff
@@ -143,22 +151,38 @@ export class App extends React.Component {
     return routes;
   }
 
-  renderExtensionTabLayoutRoutes() {
-    return clusterPageMenuRegistry.getRootItems().map((menu, index) => {
-      const tabRoutes = this.getTabLayoutRoutes(menu);
+  generateExtensionTabLayoutRoutes(rootItems: ClusterPageMenuRegistration[]) {
+    rootItems.forEach((menu, index) => {
+      let route = this.extensionRoutes.get(menu);
+
+      if (!route) {
+        const tabRoutes = this.getTabLayoutRoutes(menu);
 
-      if (tabRoutes.length > 0) {
-        const pageComponent = () => <TabLayout tabs={tabRoutes}/>;
+        if (tabRoutes.length > 0) {
+          const pageComponent = () => <TabLayout tabs={tabRoutes} />;
 
-        return <Route key={`extension-tab-layout-route-${index}`} component={pageComponent} path={tabRoutes.map((tab) => tab.routePath)}/>;
-      } else {
-        const page = clusterPageRegistry.getByPageTarget(menu.target);
+          route = <Route key={`extension-tab-layout-route-${index}`} component={pageComponent} path={tabRoutes.map((tab) => tab.routePath)} />;
+          this.extensionRoutes.set(menu, route);
+        } else {
+          const page = clusterPageRegistry.getByPageTarget(menu.target);
 
-        if (page) {
-          return <Route key={`extension-tab-layout-route-${index}`} path={page.url} component={page.components.Page}/>;
+          if (page) {
+            route = <Route key={`extension-tab-layout-route-${index}`} path={page.url} component={page.components.Page}/>;
+            this.extensionRoutes.set(menu, route);
+          }
         }
       }
     });
+
+    Array.from(this.extensionRoutes.keys()).forEach((menu) => {
+      if (!rootItems.includes(menu)) {
+        this.extensionRoutes.delete(menu);
+      }
+    });
+  }
```

Comment:
We shouldn't create an array just to iterate through it once and the throw it away.
```suggestion
    for (const meny of this.extensionRoutes.keys()) {
      if (!rootItems.includes(menu)) {
        this.extensionRoutes.delete(menu);
      }
    }
```

Reply:
I'll create a separate PR about cleanup. I never remember that we can use `for` in these cases.

Reply:
Ah I see, that is why `forEach` is so common. Though remember that `forEach` is one of the only "methods" on interators, so you don't need to convert to an array to use that method.